src/town_cmd.cpp

```diff
@@ -3243,7 +3243,7 @@ static CommandCost TownActionBuyRights(Town *t, DoCommandFlag flags)
 	/* Check if it's allowed to buy the rights */
 	if (!_settings_game.economy.exclusive_rights) return CMD_ERROR;
 
-	if (flags & DC_EXEC) {
+	if (flags & DC_EXEC && t->exclusivity == INVALID_COMPANY) {
```

Comment:
This is not right, this will cause the command to seem to succeed, despite a precondition having failed. The calling function (`CmdDoTownAction` on line 3375) will then think the action has succeeded, and subtract the cost from the company attempting the action.

You need to make a check _before_ this if statement, where you fail the command if the precondition of nobody else having exclusivity fails.

Reply:
Would that be on line 3244?

The possibility to buy the rights is disabled in the GUI when it can't be done, but I guess that doesn't stop an AI from trying.

Reply:
It's also a potential race condition if two players just happen to click the button on the exact same tick. Then one will properly succeed, but the other will get nothing but still lose their money.

Reply:
This race condition is the current behaviour. If two companies spend money on rights both will lose them but only one will walk away with the contract.

I would say that this race condition is sufficiently improbable as to not really care about, but I'll definitely fix it if someone can tell me how.

Reply:
Also, it is good to remember people make customised clients. So every check done in the GUI also nees to be done on command level, as bypassing the GUI is simple (and will be done). And we kinda want that a command does the same during testing vs execution. As a bit of background.

I think if you move the check one line up in its own if with a CMD_ERROR return, you are good already. But didn't look indepth enough to see if that causes you other trouble :)

Reply:
Fixed.

```
diff --git a/src/town_cmd.cpp b/src/town_cmd.cpp
index d6446bf27..6891504b9 100644
--- a/src/town_cmd.cpp
+++ b/src/town_cmd.cpp
@@ -3242,8 +3242,9 @@ static CommandCost TownActionBuyRights(Town *t, DoCommandFlag flags)
 {
        /* Check if it's allowed to buy the rights */
        if (!_settings_game.economy.exclusive_rights) return CMD_ERROR;
+       if (t->exclusivity != INVALID_COMPANY) return CMD_ERROR;
 
-       if (flags & DC_EXEC && t->exclusivity == INVALID_COMPANY) {
+       if (flags & DC_EXEC) {
                t->exclusive_counter = 12;
                t->exclusivity = _current_company;

```